k-distribution/k-tutorial/1_basic/05_modules/README.md

```diff
@@ -1 +1,237 @@
+include all of the sentences in the module being imported into the module that
+we import from. There are a few minor differences between importing a module
+and simply including its sentences in another module directly, but we will
+cover these differences later. Essentially, you can think of modules as
+a way of conceptually grouping sentences in a larger K definition.
+
+### Exercise
+
+Modify `lesson-05-d.k` to include four modules: one containing the syntax, two
+with one rule each that imports the first module, and one module containing
+no sentences that imports the second and third module. 
+
+## Parsing in the presence of multiple modules
+
+As you may have noticed, each module in a definition can express a distinct set
+of syntax. When parsing the sentences in a module, we use the syntax
+**of that module**, enriched with the basic syntax of K, in order to parse
+rules in that module. For example, the following definition is a parser error
+(`lesson-05-e.k`):
+
+```k
+module LESSON-05-E-1
+  rule not true => false
+  rule not false => true
+endmodule
+
+module LESSON-05-E-2
+  syntax Boolean ::= "true" | "false"
+  syntax Boolean ::= "not" Boolean [function]
+endmodule
+```
+
+This is because the syntax referenced in module `LESSON-05-E-1`, namely, `not`,
+`true`, and `false`, is not imported by that module. You can solve this problem
+ by simply importing the modules containing the syntax you want to use in your
+sentences.
+
+## Main syntax and semantics modules
+
+When we are compiling a K definition, we need to know where to start. We
+designate two specific **entry point modules**: the **main syntax module**
+and the **main semantics module**. The main syntax module, as well as all the
+modules it imports recursively, are used to create the parser for programs that
+you use to parse programs that you execute with `krun`. The main semantics
+module, as well as all the modules it imports recursively, are used to
+determine the rules that can be applied at runtime in order to execute a
+program. For example, in the above example, if the main semantics module is
+module `LESSON-05-D-1`, then `not` is an uninterpreted function and the rules
+in module `LESSON-05-D-2` are ignored.
+
+While you can specify the entry point modules explicitly by passing the
+`--main-module` and `--syntax-module` flags to `kompile`, by default, if you
+type `kompile foo.k`, then the main semantics module will be `FOO` and the
+main syntax module will be `FOO-SYNTAX`.
+
+## Splitting a definition into multiple files
+
+So far, while we have discussed ways to break definitions into multiple
+conceptual components, this is of limited usefulness because we are still
+forced to put all the modules in the definition into a single file. This is
+obviously not desirable. As a result, K provides a second mechanism for
+combining files in a K definition, namely, the **requires** directive.
+
+A `requires` statement appears at the top of a K file, prior to any module
```

Comment:
Maybe mention that there is another use for the keyword `requires`, so that someone who is looking at a K spec isn't confused. You could just mention that the meaning of that other use of `requires` will be covered later.